### ISSUE TITLE
Installation error on macbook

### ISSUE THREAD

**Issue by zhucaoxiang** [hiddenSymmetries/simsopt#139]
Hi, I am trying to install SIMSOPT on my macbook using GCC-10. I mannually set $CC and $CXX to be directed to GCC execubables. After I ran `pip install -e .[MPI,SPEC]`, no errors show up. (I will get errors if I didn't specify $CC, $CXX.)

But when I tried to import simsopt, I got the following error, which is probably caused by the `pybind`.
```
In [1]: import simsopt                                                                                                                                                                                                                                                                                              
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
RuntimeError: module compiled against API version 0xe but this version of numpy is 0xd
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-ff876a4d7849> in <module>
----> 1 import simsopt

~/Documents/Codes/simsopt/src/simsopt/__init__.py in <module>
     20 
     21 from ._core import make_optimizable
---> 22 from .objectives import LeastSquaresTerm, LeastSquaresProblem
     23 from .solve import least_squares_serial_solve
     24 from .util import initialize_logging

~/Documents/Codes/simsopt/src/simsopt/objectives/__init__.py in <module>
----> 1 from .least_squares import LeastSquaresTerm, LeastSquaresProblem
      2 
      3 __all__ = ['LeastSquaresTerm', 'LeastSquaresProblem']

~/Documents/Codes/simsopt/src/simsopt/objectives/least_squares.py in <module>
     17 from scipy.optimize import least_squares
     18 
---> 19 from .._core.dofs import Dofs
     20 from .._core.util import isnumber
     21 from .._core.optimizable import function_from_user, Target

~/Documents/Codes/simsopt/src/simsopt/_core/dofs.py in <module>
     15 
     16 from .optimizable import function_from_user
---> 17 from .util import unique, ObjectiveFailure, finite_difference_steps
     18 
     19 logger = logging.getLogger(__name__)

~/Documents/Codes/simsopt/src/simsopt/_core/util.py in <module>
     16 
     17 from ..util.types import RealArray
---> 18 from simsoptpp import Curve   # To obtain pybind11 metaclass
     19 
     20 

ImportError: SystemError: <built-in method __contains__ of dict object at 0x7ffc35657180> returned a result with an error set
```

After searching the internet, it looks there is some solutions ([link](https://github.com/kpeeters/cadabra2/issues/111)), but the commits is not simple. Has anyone saw this before?

**Comment by mbkumar** [comment]
@zhucaoxiang 
I haven't seen this issue before. If you are just trying to get simsopt installed on mac, please use this [wiki](https://github.com/hiddenSymmetries/simsopt/wiki/Mac-Installation-Instructions) where I setup simsopt on mac using anaconda tool-chain.

**Comment by zhucaoxiang** [comment]
@mbkumar Thanks for pointing out that page. I tend to keep my installation of OpenMPI using GCC. I had many problems using clang on MacBook when compiling other codes.

**Comment by mbkumar** [comment]
I was able to use the conda toolchain even with other things installed.
Perhaps you should go that route. The conda-toolchain installation
instructions are at the end.

Bharat Medasani

Engineer
Princeton Plasma Physics Lab (PPPL)

**Comment by florianwechsung** [comment]
Hi @zhucaoxiang,

Can you give a few more details on your setup? i.e. are you using the system python, or a homebrew python? And are you using a virtual environment? I have encountered issues with pybind in the past when there was an existing pybind installation that was mistakenly picked up instead of the local one. e.g. there might be one in your homebrew/macports/conda folder.

**Comment by zhucaoxiang** [comment]
@florianwechsung Good questions.

1. I am using the python executable installed from Anaconda (`/opt/anaconda3/bin/python`). 
2. I don't think I have installed pybind11 before. So by default, I was probably using the submodule in the `thirdparty/pybind11/`. That one might be installed using clang, instead of gcc. I probably should re-install pybind11. I am not an expert in cmake. Should I just remove `./build` and re-install?
3. I do notice there is an error when I do the following line. It looks weird.
```
In [1]: from pybind11.setup_helpers import Pybind11Extension                                                                                                                                                                                                                                                        
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-0f19c6a250f4> in <module>
----> 1 from pybind11.setup_helpers import Pybind11Extension

ModuleNotFoundError: No module named 'pybind11'
```

**Comment by florianwechsung** [comment]
Can you please try
```
mkdir cmake-build
cd cmake-build
env CC=gcc-10 CXX=gcc-10 cmake ..
make VERBOSE=1
```
and paste the full output of that

**Comment by zhucaoxiang** [comment]
```
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Checking whether C compiler has -isysroot
-- Checking whether C compiler has -isysroot - yes
-- Checking whether C compiler supports OSX deployment target flag
-- Checking whether C compiler supports OSX deployment target flag - yes
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/bin/gcc-10 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Checking whether CXX compiler has -isysroot
-- Checking whether CXX compiler has -isysroot - yes
-- Checking whether CXX compiler supports OSX deployment target flag
-- Checking whether CXX compiler supports OSX deployment target flag - yes
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/bin/gcc-10 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Python: /opt/anaconda3/bin/python3.8 (found suitable version "3.8.3", minimum required is "3") found components: Interpreter Development.Module NumPy
statusPython executable is /opt/anaconda3/bin/python3.8
statusPython Development Module found value is TRUE
statusPython header dirs are /opt/anaconda3/include/python3.8
statusPython library dirs are
statusPython version is 3.8.3
-- Submodule update
-- Local build detected. Set compilation flags accordingly (march=native).
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Could NOT find Boost (missing: Boost_INCLUDE_DIR) (Required is at least version "1.60.0")
-- Downloading and installing boost.
-- Boost include dirs are /Users/caoxiang/Documents/Codes/simsopt/cmake-build/thirdparty/boost/src/boost
-- pybind11 v2.6.2 dev1
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Failed
-- Version: 7.1.4
-- Build type:
-- CXX_STANDARD: 11
-- Performing Test has_std_11_flag
-- Performing Test has_std_11_flag - Success
-- Performing Test has_std_0x_flag
-- Performing Test has_std_0x_flag - Success
-- Performing Test SUPPORTS_USER_DEFINED_LITERALS
-- Performing Test SUPPORTS_USER_DEFINED_LITERALS - Success
-- Performing Test FMT_HAS_VARIANT
-- Performing Test FMT_HAS_VARIANT - Success
-- Required features: cxx_variadic_templates
-- Looking for strtod_l
-- Looking for strtod_l - found
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/caoxiang/Documents/Codes/simsopt/cmake-build
```

These lines imply some problems
```
-- Could NOT find Boost (missing: Boost_INCLUDE_DIR) (Required is at least version "1.60.0")
-- Downloading and installing boost.
-- Boost include dirs are /Users/caoxiang/Documents/Codes/simsopt/cmake-build/thirdparty/boost/src/boost
-- pybind11 v2.6.2 dev1
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Failed
```

**Comment by florianwechsung** [comment]
I don't see an obvious issue with that. What's the output of the last command I gave?

**Comment by florianwechsung** [comment]
Looking at this output again, what I find more concerning is that the C++ standard seems to be set to C++11. @mbkumar do you have an idea why our setting of C++14 is being ignored?

**Comment by zhucaoxiang** [comment]
Is `-- Performing Test HAS_FLTO - Failed` indicating some issues?

**Comment by mbkumar** [comment]
@florian,

I noticed that we are setting the language to be c++14 on the target. What
if we specify the language globally using the two commands.

set(CMAKE_CXX_STANDARD 14)set(CMAKE_CXX_STANDARD_REQUIRED ON)



Bharat Medasani

Engineer
Princeton Plasma Physics Lab (PPPL)

**Comment by florianwechsung** [comment]
@zhucaoxiang That's link time optimization and the gcc on my mac also doesn't have that.

**Comment by zhucaoxiang** [comment]
The issue was resolved.  Thank you all. @mbkumar @florianwechsung @landreman 

I tried what Matt suggested and used `python setup.py develop`. Then I found that the screen shows "it cannot find Ninja".  So I installed Ninja (with some struggles as `brew install` failed). Then I re-installed SIMSOPT using `python setup.py develop` and now it is working.

I am not quite sure if this case can be useful to improve the cmake checklist or not. But I am glad it is resolved.
